### ISSUE TITLE
[E::fai_load3_core] Failed to open FASTA file

### ISSUE THREAD

**Issue by Argonvi** [stjude/RNAIndel#25]
Hello,
Executing rnaindel PredictIndels gives me this error, with both the docker and pip installed versions. I have tried various fasta files (previously indexed with samtools faidx) as reference, but the error persists. When inspecting the running processes with "top", rnaindel seems to be running, but after leaving it for more than a day it produces no output.
I would greatly appreciate your help.

**Comment by rawagiha** [comment]
Thanks a lot for reporting the error.

Here is a similar report: 
https://github.com/google/deepvariant/issues/310

I think you have the .fai file (index) file in the same directory as the fasta file.
However, would you please try the followings?
1.  check if the fasta and fai files are indeed in the specified directory.
2.  test-run the sample data to confirm the correct installation.
     the runtime for the sample date should be < 30 sec.
     
     step1:  curl -LO https://zenodo.org/record/6109305/files/sampledataset.tar.gz
     step2:  tar -zxf sampledataset.tar.gz
     step3:  rnaindel PredictIndels -i sampledataset/sample.bam  -r sampledataset/chr10.fa -d data_dir_grch38 -o test.vcf

3.  please tell me your pysam version.

Please let me know what you get.
I appreciate your cooperation.

**Comment by Argonvi** [comment]
Thank you for your response.

1. The fasta and fai files are in the same directory.
2. The sample data was correctly processed and created the test.vcf file without any issue.
3. Re-indexed the fasta reference, to no avail.
4. I am running python 3.9.5 with pysam 0.18.0.

**Comment by rawagiha** [comment]
Thanks for checking.

Hmm..., all of 1-4 look good.
Is it okay for you to use the fasta from NCBI?

step1:
 wget ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz
step2;
gunzip GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz
step3:
samtools faidx GCA_000001405.15_GRCh38_no_alt_analysis_set.fna

If you wish to work with your version of fasta and can share the file with me, I will check it on my side.  
Please https://github.com/stjude/RNAIndel#contact me.

**Comment by Argonvi** [comment]
Hello Kohei,
I did as you suggested, but got the same error. I tried indexing my fasta reference and the one from NCBI with two versions of samtools (1.13 and 1.15.1) but it did not make a difference.
Using one of my BAM files and the reference provided in the sample dataset (chr10.fa) worked. It seems that the problem is with the fasta files I've been using.

Also, I want to point out that my error message is a bit different from that on issue [google/deepvariant/issues/310](https://github.com/google/deepvariant/issues/310), as my message says "Failed to open FASTA **file**" and theirs says "Failed to open FASTA **index**".

**Comment by rawagiha** [comment]
@Argonvi, thanks for the updates.
Based on your experiment with your BAM + chr10.fa, I agree that there may be something wrong with the fasta files.
On my side, I tested fasta files from NCBI, Ensemble and UCSC. All worked fine.

I have made my fasta (origin: NCBI) and fai files temporarily available. "chr10.fa" is the chromosome 10 sequence from this file.

https://www.transfernow.net/en/dltransfer?utm_source=20220812j9ac7A3g

Please check the file permission after you download the files.

I was not able to find error reports with the exactly same message. 
Here are similar reports
permission related:
https://github.com/katholt/genotyphi/issues/24
corrupted file:
https://github.com/atks/vt/issues/81

**Comment by Argonvi** [comment]
Hello and sorry for the late reply.

After running a couple more tests, it seems like the problem is not with the fasta files, but the bam files. I must have done something wrong in my previous experiment with my BAM and chr10.fa because I cannot reproduce the results I obtained.

After predicting indels successfully with sample.bam and my fasta reference, I tried subsetting my bam multiple times to smaller sizes with "samtools view -s", from 2.9G to 1.5G, 750M, 317M and 167M. PredictIndels failed with all bams except for the smallest (167M), and the error says "[E::fai_load3_core] Failed to open FASTA file", as the ones described in the issues you linked.

In summary, the fasta files and their indexes seem to be correct, but for some reason the bams are not. I am not completely sure if it is due to their size. Do you have any ideas?

**Comment by rawagiha** [comment]
Thanks for further testing with your data. This is extremely helpful.

We typically run the tool on 5 - 20GB BAM files here and elsewhere. So, this is not an issue with the file size.

I am guessing that some of the reads in the BAM file are problematic and that such reads are removed in the smallest file (167M) due to the most aggressive down-sampling. 

Would you please send me the second smallest one (317M)? I will only use it for debugging purpose and will delete once done.

**Comment by Argonvi** [comment]
You can download the bam from this link 

https://www.transfernow.net/dl/20220818MUTnQ6N8

**Comment by rawagiha** [comment]
Thanks for sharing the sample data with me.

It ran successfully. 
Here is the output VCF for your review (275K) : 
https://www.transfernow.net/dl/20220818kSzuNisW

Reads in the 317M BAM file appear legitimate. 

Your test-runs are only good with smaller BAM files (167M and 18M (sample.bam)). This issue may be related to memory as the consumption by RNAIndel is proportional to the input BAM file size. On my platform, the run consumed 2465 MB at peak (average: 1637 MB) to analyze your BAM file with a single core (i.e., -p 1 (default)). The run time was 277 sec (wall). Your runs have requested enough memory? Please check the memory option on your platform (**not** "-m" option with RNAIndel, this is for java heap space). 

Also, please show me the entire error message and tell me how long it takes to throw the error since the program start.

P.S.
For my run, I used the FASTA file I linked before. Does it still fail with this FASTA?

**Comment by Argonvi** [comment]
When executing rnaindel PredictIndels with "-p 1" it runs succesfully for the 317M and smaller bams, but not for the larger ones.
I am working in an ubuntu 20 server with 16 cpus and 59G of available memory.

This is the output of "top" for the process rnaindel right before the error message:
529245 agonzal+  20   0 1525008 177408  56060 R 100.0   0.3   2:43.56 rnaindel
[E::fai_load3_core] Failed to open FASTA file ../references/GRCh38.fa

If I use a smaller file (e.g. 317M) with "-p 4" I get the same error, also multiple times. The process "rnaindel" keeps running after this but it will just get stuck there indefinitely, without producing any kind of output.

**Comment by rawagiha** [comment]
Good news.
I was able to reproduce your error on my desktop. 

I have been working on the high-performance computing facility (HPCF) at my institute. On this platform, the "-p 4" does not affect the behavior. However, on my desktop (a usual Dell machine) , "-p 1" works but "-p 4" does not for the program installed via PyPI. Strangely, the docker version works with "-p 4". Also, I noticed that it is so slow on my desktop with a lot of noise from the machine.

Now that I can make your error happen, I will look into the code. In the meantime, is it possible for you to work on another machine?  

I will get back as soon as I make progress.

**Comment by rawagiha** [comment]
UPDATE.
Please try the hotfix version as follows.

1. pip uninstall rnaindel
    (please make sure that the old version is surely removed)
2. git clone -b fix https://github.com/stjude/RNAIndel.git
3. cd RNAIndel
4. python setup.py install 
    (you may need --user option: python setup.py install --user)
5. cd ..
6. rnaindel --version
   (you will see: 3.1.0)
7. rnaindel PredictIndels -i FER15_genome_10pt.bam -r GRCh38.fa -d data_dir_grch38 -o test.vcf  -p 4

**Comment by Argonvi** [comment]
The fix worked for the smaller file (317M) but not for any of the files larger than that (750M and above) when using p=4.

**Comment by rawagiha** [comment]
With p = 1, the fix works for the largest one (2.9G)? Still the same message with p > 1?

**Comment by Argonvi** [comment]
With p=1 it works even for the largest file, but p > 1 does not.

**Comment by rawagiha** [comment]
Further workaround implemented.

Please repeat the steps 1- 6 above (you will still see 3.1.0). 
When using p > 1, please append **--safety-mode**:
`rnaindel PredictIndel - i the_largest.bam -r GRCh38.fa -d data_dir_grch38 -o test.vcf -p 4 --safety-mode`

**Comment by Argonvi** [comment]
It worked! My last try was with the complete file and 8 processes and it run without problems. This should be good for processing the rest of my samples. I will let you know if I run into any more problems. Thank you for you work!

**Comment by rawagiha** [comment]
I really appreciate your patience. Thank you so much for contributing to the tool improvement.
